contracts/okp4-objectarium/src/contract.rs

```diff
@@ -118,50 +118,49 @@ pub mod execute {
         let id = crypto::hash(&bucket.config.hash_algorithm.into(), &data.0);
         let data_path = DATA.key(id.clone());
 
-        if data_path.has(deps.storage) {
-            return Err(ContractError::Bucket(BucketError::ObjectAlreadyStored));
-        }
-        let compressed_data = compression.compress(&data.0)?;
-
-        data_path.save(deps.storage, &compressed_data)?;
-
-        // store object
-        let compressed_size = (compressed_data.len() as u128).into();
-        let object = &Object {
-            id,
-            owner: info.sender.clone(),
-            size,
-            pin_count: if pin { Uint128::one() } else { Uint128::zero() },
-            compression,
-            compressed_size,
-        };
+        if !data_path.has(deps.storage) {
+            let compressed_data = compression.compress(&data.0)?;
+
+            data_path.save(deps.storage, &compressed_data)?;
+
+            // store object
+            let compressed_size = (compressed_data.len() as u128).into();
+            let object = &Object {
+                id: id.clone(),
+                owner: info.sender.clone(),
+                size,
+                pin_count: if pin { Uint128::one() } else { Uint128::zero() },
+                compression,
+                compressed_size,
+            };
 
-        objects().save(deps.storage, object.id.clone(), object)?;
+            objects().save(deps.storage, object.id.clone(), object)?;
 
-        // save bucket stats
-        BUCKET.update(deps.storage, |mut bucket| -> Result<_, ContractError> {
-            let stat = &mut bucket.stat;
-            stat.size += size;
-            stat.object_count += Uint128::one();
-            stat.compressed_size += compressed_size;
-            Ok(bucket)
-        })?;
+            // save bucket stats
+            BUCKET.update(deps.storage, |mut bucket| -> Result<_, ContractError> {
+                let stat = &mut bucket.stat;
+                stat.size += size;
+                stat.object_count += Uint128::one();
+                stat.compressed_size += compressed_size;
+                Ok(bucket)
+            })?;
+        }
 
         // save pin
         if pin {
             pins().save(
```

Comment:
Since the object might already exist and could be pinned by several addresses, saving the pin state bypasses the limitations on the total number of pins per object (`max_object_pins`). It would be appropriate to call the `pin_object()` function, which properly handles these aspects I guess.